x/ecocredit/server/marketplace/buy.go

```diff
@@ -2,10 +2,268 @@ package marketplace
 
 import (
 	"context"
+	"fmt"
 
-	marketplace "github.com/regen-network/regen-ledger/x/ecocredit/marketplace"
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/marketplace/v1"
+	ecoApi "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+	"github.com/regen-network/regen-ledger/x/ecocredit/marketplace"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/utils"
+
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+// Buy allows users to purchase credits by either directly specifying a sell order, or
+// defining a set of filters with attributes to match against.
+//
+// Currently, only the former is supported. Calls to this function with anything other than
+// MsgBuy_Order_Selection_SellOrderId will fail.
 func (k Keeper) Buy(ctx context.Context, req *marketplace.MsgBuy) (*marketplace.MsgBuyResponse, error) {
-	panic("not implemented")
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	buyerAcc, err := sdk.AccAddressFromBech32(req.Buyer)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, order := range req.Orders {
+		switch selection := order.Selection.Sum.(type) {
+		case *marketplace.MsgBuy_Order_Selection_SellOrderId:
+			sellOrder, err := k.stateStore.SellOrderTable().Get(ctx, selection.SellOrderId)
+			if err != nil {
+				return nil, fmt.Errorf("sell order %d: %w", selection.SellOrderId, err)
+			}
+			if sellOrder.DisableAutoRetire && !order.DisableAutoRetire {
```

Comment:
```suggestion
			if order.DisableAutoRetire && !sellOrder.DisableRetire {
```

it's the opposite thing we care about